packages/publisher/index.ts

```diff
@@ -0,0 +1,226 @@
+import type { Init, CommandModule, Emitter, Logging } from '@sern/handler'
+import { controller, CommandInitPlugin, CommandType } from '@sern/handler'
+import { writeFile } from 'node:fs/promises';
+import { inspect } from 'node:util';
+import type { PermissionFlagsBits } from 'discord.js'
+
+const optionsTransformer = (ops?: Array<{ type: number }>) => {
+    return ops?.map((el) => {
+        if ('command' in el) {
+            const { command, ...rest } = el;
+            return rest;
+        }
+        return el;
+    }) ?? [];
+};
+
+const intoApplicationType = (type: number) => 
+    type === 3 ? 1 : Math.log2(type);
+
+const makeDescription = (type: number, desc: string) => {
+    if (type !== CommandType.Text && desc !== '') {
+        console.warn('Found context menu that has non empty description field. Implictly publishing with empty description');
+        return '';
+    }
+    return desc;
+};
+
+const serializePerms = (perms: unknown) => {
+    if(typeof perms === 'bigint' || typeof perms === 'number') {
+       return perms.toString(); 
+    }
+
+    if(Array.isArray(perms)) {
+        return perms.reduce((acc, cur) => acc|cur, BigInt(0))
+                    .toString()
+    }
+    return null;
+}
+
+const BASE_URL = new URL('https://discord.com/api/v10/applications/');
+const PUBLISHABLE = 0b1110;
+
+export class Publisher implements Init {
+    constructor(private modules: Map<string, CommandModule>,
+                private sernEmitter : Emitter,
+                private logger: Logging) {}
+
+    async init() {
+        if(!process.env.DISCORD_TOKEN) {
+            throw Error("No token found to publish. add DISCORD_TOKEN to .env");
+        }
+        const headers = [['Authorization', 'Bot ' + process.env.DISCORD_TOKEN],
+                         ['Content-Type', 'application/json']] as Array<[string,string]>
+        let me; 
+        let appid: string;
+        try {
+            me = await fetch(new URL('@me', BASE_URL), { headers }).then(res => res.json());
+            appid = me.id;
+        } catch(e) {
+            console.log("Something went wrong while trying to fetch your application:");
+            throw e;
+        }
+        const GLOBAL_URL = new URL(`${appid}/commands`, BASE_URL);
+        interface LocalPublish {
+            guildIds?: string[]
+            default_member_permissions: string,
+            integration_types: string[],
+            contexts: number[]
+        }
+        const listener = async () => {
+            this.logger.info({ message: 'publishing modules' });
+            const modules = 
+                Array.from(this.modules.values())
+                     .filter(module => (module.type & PUBLISHABLE) != 0)
+                     .map(module => {
+                        const publish = module.locals.publish as LocalPublish || {}
```

Comment:
guarantee this to typescript that it is not gonna be a text command

Reply:
guaranteed since we can use bitwise publishable